### ISSUE TITLE
cargo check does not detect `arithmetic_overflow`

### ISSUE THREAD

**Issue by kamulos** [rust-lang/rust#116842]
I tried this code:

```rust
fn main() {
    _ = 2u32 << 32;
}
```

I expected to see this happen: the lint `arithmetic_overflow` triggering during a `check` run

Instead, this happened: the lint is only triggered when using `cargo build` or `cargo clippy`

### Meta

`rustc --version --verbose`:
```
rustc 1.75.0-nightly (49691b1f7 2023-10-16)
binary: rustc
commit-hash: 49691b1f70d71dd7b8349c332b7f277ee527bf08
commit-date: 2023-10-16
host: x86_64-unknown-linux-gnu
release: 1.75.0-nightly
LLVM version: 17.0.2
```

**Comment by ehuss** [comment]
Thanks for the report! This is a known issue where MIR-based lints are not processed in `cargo check` due to performance concerns. Closing as a duplicate of #49292. [RFC 3477](https://rust-lang.github.io/rfcs/3477-cargo-check-lang-policy.html) also contains more context.
